text/359-construct-hub-deny-list.md

```diff
@@ -0,0 +1,127 @@
+
+To add a package to the deny list, add it to the `denyList` option when defining the `ConstructHub` construct in your deployment:
+
+```ts
+new ConstructHub(this, 'ConstrucHub', {
+  denyList: {
+    '@aws-cdk/aws-eks@1.44.0': {
+      reason: 'Security issue.',
+      user: 'eladb',
+      date: new Date('2021-07-11T12:57:48.693Z'),
+      ref: 'T123333322',
+    },
+    'boombam', {
+      reason: 'Copyright violation',
+      user: 'theadmin',
+      date: new Date('2033-07-11T12:57:48.693Z'),
+      ref: 'T137838738383',
+    },
+  }
+});
+```
+
+The keys are package names and may also include specific version numbers. 
+If a version is not specified (e.g.`boombam`), then all versions of a specific module are denied. 
+The value of each entry includes metadata which will be emitted in the ingestion log and in the future displayed in the 
+"where is my package" page for publisher diagnostics.
+
+---
+
+Ticking the box below indicates that the public API of this RFC has been
+signed-off by the API bar raiser (the `api-approved` label was applied to the
+RFC pull request):
+
+```
+[ ] Signed-off by API Bar Raiser @xxxxx
+```
+
+## Public FAQ
+
+
+### What are we launching today?
+
+A feature of the Construct Hub to allow operators to block packages (or package versions) from
+appearing in search results and package pages.
+
+### Why should I use this feature?
+
+Operators (incl. Amazon, as the operator of the public construct hub) can use this to block
+packages for any reason (e.g. security, compliance, copyright, trademark, etc).
+
+## Internal FAQ
+
+### Why are we doing this?
+
+This is a security and legal requirement for the Construct Hub.
+
+### What is the technical solution (design) of this feature?
+
+The deny list will be modeled through strongly-typed API of the `ConstructHub`.
+
+1. During deployment, we will use `BucketDeployment` in order to upload this list into a file in an S3 bucket.
+2. We will trigger a lambda function every time the deny list file is created/updated. This lambda function will 
+   iterate over the packages in the deny list and will delete any objects from the packages s3 bucket that match a 
+   denied package. The object key prefix will be based on the deny entry key (name + version or just name). 
+   We currently assume this process will fit into a single lambda execution given we will issue an S3 ListObjects 
+   with a prefix filter. Timeout alarms will let us know if this does not scale and requires an additional indirection.
+3. The "discovery" and "ingestion" lambda functions will retrieve this file for each request, and will consult the 
+   deny list for every package. If an incoming package is in the deny list, it will be skipped and a log entry will 
+   be emitted with the deny information.
+4. The "inventory" process will retrieve the deny list and will filter out all denied package from inventory. 
+   Statistics about denied package will also be included in the inventory metrics.
```

Comment:
I'm not sure I understand what this means... Can you elaborate / make it more concrete?

I suppose this means the inventory function would give out a metric for how many artefacts still exist for deny-listed packages? Or does it simply give a metric for how many entries are currently in the deny-list?

Reply:
The latter - just emit the number of deny list entries. I'll make that clearer.